lib/tasks/migrate.rake

```diff
@@ -1,76 +1,226 @@
+  primary key (ID) to start migration from which can be used to resume
+  previously stopped migration. Example:
+
+  foreman-rake host_reports:migrate from_date=2021-01-01 from_id=1234567
+  END_DESC
   task :migrate => :environment do
-    ConfigReport.all.order(id: :asc).where("id >= ?", (ENV['start'] || '0').to_i).each do |r|
-      if r.origin != "Puppet"
-        next
+    Rails.logger.level = Logger::ERROR
+    Foreman::Logging.logger('permissions').level = Logger::ERROR
+    Foreman::Logging.logger('audit').level = Logger::ERROR
+    Signal.trap("INT") do
+      puts "Graceful shutdown..."
+      reports_migrate_running = false
+    end
+
+    from_id = (ENV['from_id'] || '0').to_i
+    from_date = ENV['from_date'] || '1980-01-15'
+    report_count = ConfigReport.unscoped.where("id >= ? and reported_at >= ?", from_id, from_date).count
+    print_each = 1 + (report_count / 100).to_i
+    puts "Starting, #{report_count} report(s) left"
+    counter = 0
+    ConfigReport.unscoped.all.where("id >= ? and reported_at >= ?", from_id, from_date).find_each do |r|
+      raise("Interrupted") unless reports_migrate_running
+      counter += 1
+      puts("Processing report #{counter} out of #{report_count} reports") if (counter % print_each).zero?
+      if r.origin == "Puppet"
+        logs = create_puppet_logs(r.id, r.logs)
+        keywords = detect_puppet_keywords(r.status, logs)
+      elsif r.origin == "Ansible"
+        results = create_ansible_results(r.id, r.logs)
+        keywords = detect_ansible_keywords(r.status)
+      end
+      if keywords.present?
+        keywords_to_insert = keywords.each_with_object([]) do |n, ks|
+          ks << { name: n }
+        end
+        ReportKeyword.upsert_all(keywords_to_insert, unique_by: :name)
+        report_keyword_ids = ReportKeyword.where(name: keywords).distinct.pluck(:id)
+      end
+      summary = summary(r.origin, r.metrics, r.status)
+      body = create_body(r.origin&.downcase, r.metrics, r.reported_at, r.status, r.host, report_keyword_ids, summary)
+      if r.origin == "Puppet"
+        body.merge!(logs: logs)
+      elsif r.origin == "Ansible"
+        body.merge!(results: results)
       end
-      body = create_body(r.metrics, r.reported_at, r.status, r.logs, r.host)
       origin = r.origin.downcase
-      host_report = host_report_create(r.host_id, origin, body)
       User.without_auditing do
         User.as_anonymous_admin do
-          HostReport.create(host_report)
+          HostReport.create!(build_report(r.host_id, origin, body, report_keyword_ids))
         end
       end
     rescue StandardError => e
-      puts "An error occurred, to resume the process: foreman-rake host_reports:migrate from=#{r.reported_at} start=#{r.id}"
-      throw e
+      puts "Error when processing report ID=#{r.id}"
+      puts r.inspect
+      puts "To resume the process:\n\n***\n\nforeman-rake host_reports:migrate from_id=#{r.id} from_date=#{from_date}\n\n***\n\n"
```

Comment:
The search does greater-or-equal-than comparison for both ID and date. Wouldn't that mean that if there is an error while exporting a report, the suggested command would then start with the report which failed to migrate, possibly just hitting the same error as previously and failing straight away?

Reply:
Yeah my idea is - you hit an error, you report to us, we fix the problem and after patching the instance you continue with the migration.

Reply:
Then it is fine